File: Solutions/LeetCode/0268/MissingNumber.java
// LeetCode #268
// https://leetcode.com/problems/missing-number/
//
// Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.

// Time: O(n)
// Space: O(1)
class Solution {
    public int missingNumber(int[] nums) {
        // Input validation.
        if (nums == null) {
            return -1;
        }

        // Perform cyclic sort.
        int i = 0;
        while (i < nums.length) {
            if (nums[i] < nums.length && nums[i] != i) {
                int temp = nums[nums[i]];
                nums[nums[i]] = nums[i];
                nums[i] = temp;
            } else {
                i++;
            }
        }

        // Find the missing number.
        for (i = 0; i < nums.length; i++) {
            if (nums[i] != i) {
                return i;
            }
        }

        // Covers the edge case where input is already in order.
        // Ex: [0, 1]
        return nums.length;
    }
}

File: solutions/LeetCode/0268/MissingNumber.cs
// LeetCode #268
// https://leetcode.com/problems/missing-number/
//
// Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.

// This solution is more optimal and uses only O(1) space.
// Time: O(n)
// Space: O(1)
public class Solution {
    public int MissingNumber(int[] nums) {
        int numsSum = 0;
        for (int i = 0; i < nums.Length; i++) {
            numsSum += nums[i];
        }
        
        int n = nums.Length + 1;
        int gaussSum = n * (n - 1) / 2;
        
        return gaussSum - numsSum;
    }
}

// This solution uses O(n) memory instead of O(1) but doesn't rely on having to know the Gauss Sum formula.
// Time: O(n)
// Space: O(n)
public class Solution {
    public int MissingNumber(int[] nums) {
        int[] foundNums = new int[nums.Length + 1];
        
        for (int i = 0; i < nums.Length; i++) {
            foundNums[nums[i]] = -1;
        }
        
        for (int i = 0; i < foundNums.Length; i++) {
            if (foundNums[i] != -1) {
                return i;
            }
        }
        
        return -1;
    }
}